src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -580,19 +580,19 @@ private static string JoinCore(ReadOnlySpan<char> separator, string?[] value, in
 
         public static string Join(string? separator, IEnumerable<string?> values)
         {
-            if (values is List<string?> valuesList)
+            if (values is null)
```

Comment:
You're right about the guideline. But in this case, the JIT needs to emit a null check anyway so that `values is List<string?>` doesn't null ref. So this null check should be free from a codegen perspective. Even better, by allowing the JIT to prove very early on in this method that the arg cannot be null, it allows null checks to be elided for the remainder of the method, which should further improve codegen.